### PR TITLE
AirfRANS: 3L/256d + EMA=0.999 (depth + EMA compound)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Your previous experiment (#3099, 3L/256d no-EMA) showed vol MSE 0.004851 — a 36% improvement over the old no-EMA baseline but worse than the new EMA baseline (0.002777). The key question: does 3L depth + EMA outperform 2L + EMA?

EMA=0.999 just transformed the 2L champion (surface 0.000459, vol 0.002777 — PR #3050). The next hypothesis is whether adding depth (3L) on top of EMA further improves volume without degrading surface.

## Instructions

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group af-3l-ema-compound
```

Report BOTH: val_primary/surface_mse AND val_primary/volume_mse.

## Baseline (UPDATED — PR #3050)
- **val_primary/surface_mse:** 0.000459 (2L + EMA=0.999)
- **val_primary/volume_mse:** 0.002777 (2L + EMA=0.999)
- **SpiderSolver targets:** Surf=0.0043, Vol=0.0017

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT spike:

## Results

### AirfRANS: 3L/256d + EMA=0.999 (depth + EMA compound)

**Hypothesis:** Does 3L depth + EMA outperform 2L + EMA?

**Verdict: FALSIFIED.** 3L+EMA is significantly worse than 2L+EMA on both metrics and is unstable — the model diverges catastrophically around epoch 135-148.

| Metric | Baseline (2L+EMA) | This run (3L+EMA best) | Δ | SpiderSolver |
|---|---|---|---|---|
| val_primary/surface_mse | 0.000459 (ep771) | 0.00288 (ep130) | +528% (6.3x worse) | 0.0043 |
| full_val/volume_mse | 0.002777 (ep771) | 0.01163 (ep129) | +319% (4.2x worse) | 0.0017 |

### Key observations

1. **Smooth descent ep1-130:** Both metrics improved monotonically through 2.5 cosine cycles. Best surface 0.00288 at ep130, best volume 0.01163 at ep129. The EMA warmup completed by ep25 (decay reached target 0.999).

2. **Catastrophic divergence ep135-149:** At the third cosine cycle LR peak (~ep125), the online model was perturbed into an unstable region. Unlike the 2L model which recovers from these perturbations, the 3L model amplified the instability. Metrics degraded: 0.003→0.007→0.023→0.344→0.621 (surface) and 0.012→0.032→0.076→0.442→1.533 (volume). Loss reached NaN at ep149.

3. **3L depth is harmful for AirfRANS:** The extra layer adds optimization noise that destabilizes training under cosine cycling, even with EMA smoothing. This is consistent with PR #2771 (3L/256d no-EMA diverged at ep202) and PR #2828 (2L/256d achieved 0.001236 vs 3L/256d at 0.001479). Depth reduction 3L→2L was a key insight for AirfRANS.

4. **EMA does NOT prevent divergence at 3L:** While EMA=0.999 preserves the best checkpoint, it cannot prevent the online model's divergence from propagating. At decay=0.999, the EMA follows ~1000 updates behind — once the online model enters a bad basin for sustained epochs, the EMA follows.

### Comparison to prior 3L results
- PR #3099 (3L/256d no-EMA): vol MSE 0.004851 — **better volume** than this run's 0.01163, suggesting the T_max=50 cycling at 3L is specifically harmful
- PR #2771 (3L/256d golden): surface MSE 0.001479 (diverged ep202) — similar divergence pattern

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --agent spike --wandb-name "spike/af-3l-ema-compound" \
  --wandb-group af-3l-ema-compound
```

- **W&B run:** ooxb9roq
- **Peak VRAM:** ~43 GB (GPU 0)
- **Epochs completed:** 149 (NaN at ep149, diverged from ep135)
- **Best checkpoint epoch:** 130 (surface), 129 (volume)

### Suggested follow-ups

1. **3L + EMA with T_max=100 or T_max=200:** The divergence is triggered by the cosine LR peak. A longer cycle with a gentler peak might allow 3L to train stably.
2. **3L + EMA + gc=0.5:** Tighter gradient clipping at the cosine peaks could prevent the perturbation that triggers divergence.
3. **3L + EMA + lower LR (3e-4 or 4e-4):** Reducing the peak LR would reduce the magnitude of the cosine peak perturbation.
4. **Conclusion: 2L is the correct depth for AirfRANS.** The evidence across multiple experiments (PRs #2771, #2828, #3099, #3139) consistently shows 2L outperforms 3L on AirfRANS. Future AirfRANS improvements should focus on the 2L architecture.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3139 because ADVISOR: surface=0.002863 (6.2x worse) and vol=0.011626 (4.2x worse) vs the EMA champion (0.000459/0.002777). Diverged catastrophically at ep149. This is consistent with all prior 3L tests on AirfRANS (PRs #2771, #2828, #3099) — 3L is harmful regardless of EMA. The extra layer adds optimization noise that destabilizes the trajectory. 2L is the confirmed optimal depth for AirfRANS. Closing — 3L direction fully dead..
